extensions/binding/org.eclipse.smarthome.binding.tradfri/src/main/java/org/eclipse/smarthome/binding/tradfri/handler/TradfriGatewayHandler.java

```diff
@@ -77,40 +85,129 @@ public void handleCommand(ChannelUID channelUID, Command command) {
+            builder.setPskStore(new StaticPskStore("Client_identity", configuration.code.getBytes()));
+
+            DTLSConnector dtlsConnector = new DTLSConnector(builder.build());
+            CoapEndpoint authEndpoint = new CoapEndpoint(dtlsConnector, NetworkConfig.getStandard());
+            authUrl = "coaps://" + configuration.host + ":" + configuration.port + "/15011/9063";
+
+            CoapClient deviceClient = new CoapClient(new URI(authUrl));
+            deviceClient.setEndpoint(authEndpoint);
+
+            JsonObject json = new JsonObject();
+            json.addProperty(CLIENT_IDENTITY_PROPOSED, identity);
+
+            gatewayResponse = deviceClient.post(json.toString(), 0);
+
+            authEndpoint.destroy();
+            deviceClient.shutdown();
+
+            if (gatewayResponse.isSuccess()) {
+                responseText = gatewayResponse.getResponseText();
+                json = new JsonParser().parse(responseText).getAsJsonObject();
+                preSharedKey = json.get(NEW_PSK_BY_GW).getAsString();
+
+                if (isNullOrEmpty(preSharedKey)) {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Pre-shared key was not obtain successfully");
+                } else {
+                    logger.info("Recieved pre-shared key for gateway '{}'", configuration.host);
+                    logger.debug("Using identity '{}' with pre-shared key '{}'. Code can be discarded now", identity, preSharedKey);
+
+                    Configuration editedConfig = editConfiguration();
+                    editedConfig.put(TradfriBindingConstants.GATEWAY_CONFIG_CODE, null);
+                    editedConfig.put(TradfriBindingConstants.GATEWAY_CONFIG_IDENTITY, identity);
+                    editedConfig.put(TradfriBindingConstants.GATEWAY_CONFIG_PRE_SHARED_KEY, preSharedKey);
+                    updateConfiguration(editedConfig);
+
+                    setupCoapClient();
+                }
+            } else {
+                logger.warn("Failed obtaining pre-shared key for identity '{}' (response code '{}', response text '{}')",
+                        identity, gatewayResponse.getCode(),
+                        isNullOrEmpty(gatewayResponse.getResponseText()) ? "<empty>" : gatewayResponse.getResponseText());
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                        String.format("Failed obtaining pre-shared key with status code '%s'", gatewayResponse.getCode()));
+            }
+        } catch (URISyntaxException e) {
+            logger.error("Illegal gateway URI '{}'", authUrl, e);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+        } catch (JsonParseException e) {
+            logger.warn("Invalid response recieved from gateway '{}'", responseText, e);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
```

Comment:
Not a configuration error

Reply:
Well - I copied the behavior from you:
```
try {
            URI uri = new URI(gatewayURI);
            deviceClient = new TradfriCoapClient(uri);
        } catch (URISyntaxException e) {
            logger.error("Illegal gateway URI '{}': {}", gatewayURI, e.getMessage());
            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
            return;
        }
```

I would agree that the user can't do anything - but what would then be the status detail - how about: ```HANDLER_INITIALIZING_ERROR``` ?

Reply:
I assume you wanted to write this comment above in the URISyntaxException catch clause.
Checking it again, the URI can indeed only be invalid if either host or port are invalid in the config, so I removed the comment above again.

In this case here (JsonParseException), I would clearly suggest a COMMUNICATION_ERROR.